packages/vhd-lib/src/Vhd/VhdAbstract.js

```diff
@@ -0,0 +1,198 @@
+  }
+
+  /**
+   * ensure the bat size can store at least entries block
+   * move blocks if needed
+   * @param {number} entries
+   */
+  ensureBatSize(entries) {
+    throw new Error(`ensuring batSize can store at least  ${entries} is not implemented`)
+  }
+
+  // Write a context footer. (At the end and beginning of a vhd file.)
+  writeFooter(onlyEndFooter = false) {
+    throw new Error(`writing footer   ${onlyEndFooter ? 'only at end' : 'on both side'} is not implemented`)
+  }
+
+  writeHeader() {
+    throw new Error(`writing header is not implemented`)
+  }
+
+  setUniqueParentLocator(fileNameString) {
+    throw new Error(`setting unique parent locator from file name ${fileNameString} is not implemented`)
+  }
+
+  _writeParentLocator(parentLocatorId, platformDataOffset, data) {
+    throw new Error(`write Parent locator ${parentLocatorId} is not implemented`)
+  }
+  // common
+  get batSize() {
+    return computeBatSize(this.header.maxTableEntries)
+  }
+
+  // return the first sector (bitmap) of a block
+  _getBatEntry(blockId) {
+    const i = blockId * 4
+    const { blockTable } = this
+    return i < blockTable.length ? blockTable.readUInt32BE(i) : BLOCK_UNUSED
+  }
+
+  // Returns the first address after metadata. (In bytes)
+  _getEndOfHeaders() {
+    const { header } = this
+
+    let end = FOOTER_SIZE + HEADER_SIZE
+
+    // Max(end, block allocation table end)
+    end = Math.max(end, header.tableOffset + this.batSize)
+
+    for (let i = 0; i < PARENT_LOCATOR_ENTRIES; i++) {
+      const entry = header.parentLocatorEntry[i]
+
+      if (entry.platformCode !== PLATFORM_NONE) {
+        end = Math.max(end, entry.platformDataOffset + sectorsToBytes(entry.platformDataSpace))
+      }
+    }
+
+    debug(`End of headers: ${end}.`)
+
+    return end
+  }
+
+  // Returns the first sector after data.
+  _getEndOfData() {
+    let end = Math.ceil(this._getEndOfHeaders() / SECTOR_SIZE)
+
+    const sectorsOfFullBlock = this.sectorsOfBitmap + this.sectorsPerBlock
+    const { maxTableEntries } = this.header
+    for (let i = 0; i < maxTableEntries; i++) {
+      const blockAddr = this._getBatEntry(i)
+
+      if (blockAddr !== BLOCK_UNUSED) {
+        end = Math.max(end, blockAddr + sectorsOfFullBlock)
+      }
+    }
+
+    debug(`End of data: ${end}.`)
+
+    return sectorsToBytes(end)
+  }
+
+  writeParentLocator(parentLocatorId, data) {
+    assert(parentLocatorId >= 0, 'parent Locator id must be a positive number')
+    assert(parentLocatorId < 8, 'parent Locator id  must be less than 8')
+    assert.notStrictEqual(this.header, undefined, `header must be read before it's used`)
+
+    let { platformDataOffset, platformDataSpace } = this.header.parentLocatorEntry[parentLocatorId]
+    // ensure there is enough place to write it
+    if (platformDataSpace > data.length) {
+      // growing : put it at the end of the data
+      platformDataOffset = this._getEndOfData()
```

Comment:
This seems specific to `VhdFile`.

Reply:
the header contains the address of the parent locator entries. We can't save the header ( and sign it ) without having an address, as opposed to the BAT where we can postpone the address computation